(DOCSP-16262): Set Up a CI/CD Pipeline (#1344)
mongodb/docs-realm: source/deploy/sdlc.txt

tasks and actions you might perform within your pipelines to configure and test
your Realm applications.

.. tip:: See a real example with GitHub Actions
   
   If you'd like to see an example CI/CD pipeline that manages testing,
   deployment, and other tasks for a real application, check out the article
   `How to Build CI/CD Pipelines for MongoDB Realm Apps Using GitHub Actions
   <https://www.mongodb.com/developer/how-to/build-ci-cd-pipelines-realm-apps-github-actions/>`__
   on the MongoDB Developer Hub.

Pipeline Stages
---------------

No two CI/CD pipelines are the same, but at a high level most pipelines share a
common pattern of moving through multiple stages that each handle different
concerns.

Development
~~~~~~~~~~~

The Development stage is the first step for creating new features and fixing
bugs in a Realm app. In this stage you work with your application's
configuration files and source code to implement your desired changes. 

To develop new features for an existing app, you can use the following high
level procedure:

1. Fork the main app and deploy a new development copy. This instance will have

>    a different App ID than your production app. Further, you can use

Reviewer: Second time using stuffier "Further":
```suggestion
   a different App ID than your production app. You can also use
```

   :ref:`environment value templates <config-expansions>` to use development
   data sources and other services that are not linked to production.

2. Develop your application. This could involve updating or adding client app
   screen, adding a new database trigger, or any other application features.

3. Run automated tests locally to ensure that your code does not introduce any
   new errors. Tests that pass locally do not guarantee that your app is free of

>    integration bugs, but indicate that your changes do not include any

Reviewer: We could hedge this even more:
```suggestion
   integration bugs, but increase confidence that your changes do not include any
```

   regressions or unintended behavior.

Staging
~~~~~~~

The Staging stage, which you might also call QA (Quality Assurance), Testing, or
Pre-Production, is a step that simulates your development changes in an
environment that is as similar to production as possible. This gives you a
usable version of your app for review and can help you catch integration bugs
with live services without affecting production data.

The specifics of your staging deployment depend on your application's needs.
However, you can use the following high level procedure to set it up:

> 1. Set up your staging environment. Use separate, non-production services and

Author: It's up to them and we don't have a strong recommendation afaict. Lauren's workflow has a consistent staging app but there's no common requirement for that to be the case.

I updated the wording here a bit, hopefully it clarifies a bit.

   possible. For example, you might use an Atlas cluster named ``staging`` that
   otherwise has the same configuration as your ``production`` cluster.

2. Create a staging build. This is best handled automatically as part of your
   CI/CD process when you create a new new pull request, but you could also
   manually build your staging environment.

3. Verify that your app behaves as expected. This might involve running an
   automated test suite against your staging environment, manually checking
   behavior, or getting approval through a user-acceptance test.

Production
~~~~~~~~~~

The Production stage is the final deployment step where your modified app is
deployed into your production environment. Ideally at this stage you have
already tested your changes locally and in staging to confirm that they are safe
to deploy. You can either deploy to production manually or automatically as part
of your CI/CD workflow.

Build Tasks
-----------

This section outlines common tasks that you will perform in your CI/CD pipeline.
You may not always do all of these tasks depending on your use case and pipeline
stage, but in general most pipelines will perform all of these at least once.

> Configure the Environment

Author: We talk about them in the development section briefly. This section is specifically focused on build script environment rather than Realm app env values, which is an unfortunate name clash. That said, maybe we can encourage them to specify their app env as a bash env variable.

~~~~~~~~~~~~~~~~~~~~~~~~~

The configuration and code for your app should generally be similar between
development stages. However, you'll want to change the value of certain
configuration options depending on the environment.

Determine what stage you're building and set the appropriate configuration
values. For example, you might configure the app with the App ID of a new app in
the Development stage or use your production App ID in the Production stage.

.. code-block:: bash
   
   # Use the production App ID for the main branch
   export REALM_APP_ID="myapp-abcde"
   # Use a staging App ID for the QA branch
   export REALM_APP_ID="myapp-staging-fghij"
   # Use a new App ID for development branches - you'll need to create the app first!
   export REALM_APP_ID="myapp-dev-zyxwv"

Set Up Realm CLI
~~~~~~~~~~~~~~~~

> Realm CLI is the easiest way to programmatically create, configure, and manage

Reviewer 2: Should this talk about or link to [Realm CLI](https://docs.mongodb.com/realm/cli/)?

> Realm apps. You should install and use the latest version in your deployment

Reviewer: Could we use a link to install here?

scripts.

You'll also need a MongoDB Atlas public/private API key pair to authenticate and
use the CLI. For more information and a walkthrough of how to get an API key,
see :atlas:`Programmatic API Keys
</configure-api-access/#programmatic-api-keys>`.

.. code-block:: bash
   
   export ATLAS_PUBLIC_KEY="<REALM_API_PUBLIC_KEY>"
   export ATLAS_PRIVATE_KEY="<REALM_API_PRIVATE_KEY>"

.. code-block:: bash
   
   npm install -g mongodb-realm-cli

>    realm-cli login --api-key=$ATLAS_PUBLIC_KEY --private-api-key=$ATLAS_PRIVATE_KEY

Reviewer: Wouldn't it be better to use the 'profiles' feature here, so you aren't storing your private keys in a script?

Author: Good call, updated to have that

   integration bugs, but increase confidence that your changes do not include any
   regressions or unintended behavior.

Staging
~~~~~~~

The Staging stage, which you might also call QA (Quality Assurance), Testing, or
Pre-Production, is a step that simulates your development changes in an
environment that is as similar to production as possible. This gives you a
usable version of your app for review and can help you catch integration bugs
with live services without affecting production data.

The specifics of your staging deployment depend on your application's needs.
However, you can use the following high level procedure to set it up:

1. **Set up your staging environment.** Use separate, non-production services and
   data sources with configurations that mirror production as closely as
   possible. For example, you might use an Atlas cluster named ``staging`` that
   otherwise has the same configuration as your ``production`` cluster.
   Depending on your use case, you may have a consistent Realm app that you
   reuse for all staging builds or you may create a new Realm app for each
   staging build.

> 2. **Create or use an existing a staging build.** You can automatically create a

Reviewer: ```suggestion
2. **Create or use an existing staging build.** You can automatically create a
```
